// websocket_p2pnet/utils/Logger.js
/**
 * 日志记录
 */
class Logger {

  /**
   * 获取当前服务器日期时间 字符串
   */
  getCurDateTimeStr() { 
    var date = new Date();
    return date.toLocaleDateString() + ' ' + date.toLocaleTimeString();
  }

  /**
  * 服务端 标题
  * @param msg 
  */
  logSTitle(funcName, msg, msg2) {
    console.log('【Server.' + funcName + '】- ' + this.getCurDateTimeStr()+'**********');
    console.log(msg);

    if (msg2) {
      console.log(msg2);
    }
    console.log('\n');
  }

  /**
   * 服务端 内容
   * @param msg 
   */
  logSContent(funcName, msg, msg2) {
    console.log('【Server.' + funcName + '】 - ' + this.getCurDateTimeStr());
    console.log(msg);

    if (msg2) {
      console.log(msg2);
    }
    console.log('\n');
  }
  
  /**
   * 客户端 标题
   * @param msg 
   */
  logCTitle(funcName, msg, msg2) {
    console.log('【Client.' + funcName + '】- ' + this.getCurDateTimeStr()+'**********');
    console.log(msg);

    if (msg2) {
      console.log(msg2);
    }
    console.log('\n');
  }

  /**
   * 客户端 内容
   * @param msg 
   */
  logCContent(funcName, msg, msg2) {
    console.log('【Client.' + funcName + '】 - ' + this.getCurDateTimeStr());
    console.log(msg);

    if (msg2) {
      console.log(msg2);
    }
    console.log('\n');
  }

 
}
module.exports = Logger;
